riscv.c

```diff
@@ -770,8 +791,485 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
 #define op_nmsub NULL
 #define op_nmadd NULL
 
+#ifdef ENABLE_RV32C
+static bool c_op_addi(struct riscv_t *rv, uint16_t inst)
+{
+    uint16_t tmp =
+        (uint16_t)(((inst & FCI_IMM_12) >> 5) | (inst & FCI_IMM_6_2)) >> 2;
+    const int32_t imm = (0x20 & tmp) ? 0xffffffc0 | tmp : tmp;
+    const uint16_t rd = c_dec_rd(inst);
+
+    // dispatch operation type
+    if (rd != 0) {
+        // C.ADDI
+        debug_print("Entered c.addi");
+        rv->X[rd] += imm;
+    } else {
+        debug_print("Entered c.nop");
+        // C.NOP
+    }
+
+    // step over instruction
+    rv->PC += rv->inst_len;
+    // enforce zero register
+    if (rd == rv_reg_zero)
+        rv->X[rv_reg_zero] = 0;
+    return true;
+}
+
+static bool c_op_addi4spn(struct riscv_t *rv, uint16_t inst)
+{
+    debug_print("Entered c.addi4spn");
```

Comment:
You can use GDB script to locate the instruction handlers instead of dumping the craft.